java/core/src/java/org/apache/orc/impl/ParserUtils.java

```diff
@@ -249,75 +255,210 @@ public static TypeDescription findSubtype(TypeDescription schema,
         case STRUCT: {
-          int posn = -1;
-          if (isSchemaEvolutionCaseAware) {
-            posn = current.getFieldNames().indexOf(first);
-          } else {
-            // Case-insensitive search like ORC 1.5
-            for (int i = 0; i < current.getFieldNames().size(); i++) {
-              if (current.getFieldNames().get(i).equalsIgnoreCase(first)) {
-                posn = i;
-                break;
-              }
-            }
-          }
-          if (posn == -1) {
-            throw new IllegalArgumentException("Field " + first +
-                " not found in " + current.toString());
-          }
-          current = current.getChildren().get(posn);
+          posn = isSchemaEvolutionCaseAware
+              ? current.getFieldNames().indexOf(first)
+              : findCaseInsensitive(current.getFieldNames(), first);
           break;
         }
         case LIST:
           if (first.equals("_elem")) {
-            current = current.getChildren().get(0);
+            posn = 0;
           } else {
-            throw new IllegalArgumentException("Field " + first +
-                "not found in " + current.toString());
+            posn = -1;
           }
           break;
         case MAP:
           if (first.equals("_key")) {
-            current = current.getChildren().get(0);
+            posn = 0;
           } else if (first.equals("_value")) {
-            current = current.getChildren().get(1);
+            posn = 1;
           } else {
-            throw new IllegalArgumentException("Field " + first +
-                "not found in " + current.toString());
+            posn = -1;
           }
           break;
         case UNION: {
           try {
-            int posn = Integer.parseInt(first);
+            posn = Integer.parseInt(first);
             if (posn < 0 || posn >= current.getChildren().size()) {
               throw new NumberFormatException("off end of union");
             }
-            current = current.getChildren().get(posn);
           } catch (NumberFormatException e) {
             throw new IllegalArgumentException("Field " + first +
                 "not found in " + current.toString(), e);
           }
           break;
         }
         default:
-          throw new IllegalArgumentException("Field " + first +
-              "not found in " + current.toString());
+          posn = -1;
+      }
+      if (posn < 0) {
+        throw new IllegalArgumentException("Field " + first +
+                                           " not found in " + current.toString());
+      }
+      current = current.getChildren().get(posn);
+      visitor.visit(current, posn);
+    }
+  }
+
+  static class ColumnFinder implements TypeVisitor {
+    private ColumnVector[] top;
+    private ColumnVector current = null;
+    private final ColumnVector[] result;
+    private int resultIdx = 0;
+
+    ColumnFinder(TypeDescription schema, VectorizedRowBatch batch, List<String> levels) {
```

Comment:
Instead of passing in the List of strings, I think I'd pass in the levels as an int.

Reply:
done